### ISSUE TITLE
get events from python

### ISSUE THREAD

**Issue by tsaridas** [saltstack/salt#37018]
### Description of Issue/Question

Trying to follow the guide at to get events from a python scripts using the salt.utils.event class,
I noticed that when master is restarted the events do not keep coming any more.
[Saltstack Documentation ](http://docs.saltstack.com/en/latest/topics/event/index.html)

I noticed @DmitryKuzmenko did some changes for the minions and syndics in order to connect after disconnection. Maybe he could suggest something similar for the specific case ?

I also tried using iter_events with the same results.

Also get_event_block in a while loop throughs an exception but when I try to reconnect It fails.
### Steps to Reproduce Issue

Follow the steps at the documentation and restart master.
### Versions Report

latest 2016.3.3

**Comment by Ch3LL** [comment]
yes I believe @DmitryKuzmenko would definitely be the one with the definitive answer.

But I would think the reason monitoring events with the master stops after restarting the master is because you would  need to re-inialize the connection to the master. The same thing happens if I were to run `salt-run state.event pretty=True` if I restart the master I have to re-run that command to view events but i think dmitry will have more knowledge around this and might have a workaround/suggestion. Thanks

**Comment by DmitryKuzmenko** [comment]
@Ch3LL is right. When master restarts it recreates the unix domain socket PUB socket. So after master restart client continues to listen the removed socket file while master publishes to the new socket file. But I'll take a closer look to provide a way to handle this.

**Comment by tsaridas** [comment]
thanks @DmitryKuzmenko,

I understand that is the same socket file and it would be nice if I had a method checking the connection or maybe automatically connect back. 
connect_pub() returns True after restart but it still doesn't show any events after reconnecting.

I tried to see where the exactly does this get stuck but I wasn't able to. 

ps: I already added your patch for syndic reconnection thinking that it might help but no luck.

**Comment by DmitryKuzmenko** [comment]
@tsaridas 
After thinking slightly more I've got that actually when publisher closes the socket all the clients get `StreamClosedError`. And the mentioned Syndic fix introduces this error handling. After looking up the code I've got that we have (now) the way to handle the error in async mode like syndic does but we have no way to detect disconnection in sync client mode, that is described in the docs. It's because for the user disconnect looks similar to timeout because of this:
http://github.com/saltstack/salt/blob/develop/salt/utils/event.py#L532-L533

So client code gets None in both cases: timeout and error. We have to fix this.

**Comment by tsaridas** [comment]
@DmitryKuzmenko thanks for checking.

Would you be able to provide a quick fix so that my client connects automatically ?

For some reason even when I try to get the exception from get_event_noblock() I cannot reconnect anymore.

Thanks

**Comment by DmitryKuzmenko** [comment]
@tsaridas ah, sure, there are more special `get_event_noblock()` and `get_event_block()` that don't hide the exception. I've tried to do things you want and everything looks working:

```python
#!/bin/env python 
import salt.config
import salt.utils.event
import sys
import time
import tornado

opts = salt.config.client_config('/etc/salt/master')

event = salt.utils.event.get_event(
        'master',
        sock_dir=opts['sock_dir'],
        transport=opts['transport'],
        opts=opts)

while True:
    try:
        while True:
            data = event.get_event_block()
            print(data)
    except tornado.iostream.StreamClosedError as ex:
        print(ex)
        event.close_pub()
        tries = 0
        while not event.cpub:
            if tries > 0:
                time.sleep(1)
            tries += 1
            print('Reconnecting #{0}'.format(tries))
            event.connect_pub()
    except:
        print('Some exception: {0}'.format(sys.exc_info()[0]))
        break

print('Done')
```

**Comment by tsaridas** [comment]
@DmitryKuzmenko ,

that doesn't work for me. It gets stuck, as I expected, at event.connect_pub() and never times out.

I also tried setting event.connect_pub(timeout=10) but without success.

I see that you are working with the develop branch hence the issue might be fixed there. 

Did you actually try to restart master when running the script ?

Thanks

**Comment by tsaridas** [comment]
Tested and it works in develop branch. Not sure which ticket solved this though.

anyway, thanks for the help @DmitryKuzmenko

**Comment by szjur** [comment]
@DmitryKuzmenko, any idea how to make it work in 2016.3.3? There is no close_pub() there, I tried to do what it does based on the latest code but it always hangs on connect_pub() even if you wait a minute and the master is surely up again at that time.

**Comment by DmitryKuzmenko** [comment]
@szjur you can either re-create the SaltEvent or use the following code instead of missing `close_pub()`:
```python
        event.subscriber.close()
        event.subscriber = None
        event.pending_events = []
        event.cpub = False
```

**Comment by tsaridas** [comment]
@Ch3LL @DmitryKuzmenko thank you for your help.
Issue can be closed.

**Comment by szjur** [comment]
@DmitryKuzmenko you do realise that what you suggested is exactly what I said tried? That doesn't help, at least not with 2016.3.3 on RHEL6. @Ch3LL,  I don't know why @tsaridas closed it - most likely he found the code change that fixes this bug but just closing it light that doesn't really help the stable branches to get better and may be a bit confusing to anyone stumbling on this bug and reading this thread.

**Comment by DmitryKuzmenko** [comment]
@szjur ah! I see, sorry. I'll take a closer look.

**Comment by tsaridas** [comment]
I could reopen the ticket if you think it should correspond to the stable version being buggy.

fyi- the pr that fixed the issue in 2016.3.3 was #32329 

Not sure why though since the specific one has to do with serializer ... I only patched ipc.py and transport/frame.py together with your syndic patch.

I will take another look at the changes one more time and try to see what was the exact line that fixed it.

Thnx

**Comment by DmitryKuzmenko** [comment]
I agree with @szjur. Also I've checked the latest 2016.3 branch and the bug is there: SaltEvent hangs on reconnecting to the event bus.
The issue is fixed in carbon branch by this PR: https://github.com/saltstack/salt/pull/36720
I think it's needed to be backported into 2016.3 branch.

**Comment by DmitryKuzmenko** [comment]
I've backported the fix to 2016.3. @cachedout could you please take a look?

**Comment by DmitryKuzmenko** [comment]
@szjur, @tsaridas thank you for help!

**Comment by szjur** [comment]
Excellent, thanks a lot @DmitryKuzmenko :-) We'll also test it thoroughly tomorrow.

**Comment by DmitryKuzmenko** [comment]
I'd like to address this question to @cachedout and @thatch45.
In simple words: `salt.utils.event` has different error handling in function allowing to get events.
`get_event_block`, `get_event_noblock`: return `None` if there's no data, raises `StreamClosedError` if remote peer is disconnected.
`get_event()`: returns None if there's no data or error occurred, we have no way to detect errors.
`iter_events()`: continues waiting for events if got None from `get_event()`, i.e. it hangs if connection is broken.
Shouldn't we change `get_event()` and `iter_events()` behavior?

**Comment by thatch45** [comment]
This is a good question, I think the optimal situation would be to allow the user to pass the option into `get_event()` to expose errors raised

**Comment by szjur** [comment]
In any case iter_events() hanging indefinitely after master restart with no way whatsoever to catch it and handle is beyond any user's expectation. It should at least return None.

**Comment by jeanpralo** [comment]
I am just wondering why should we be passing an option to raise an exception  ? I mean if it does not raise anything, this is sort of broken by default no  ?

Anyway all in favor of fixing it for get_event and so for iter_events as well.

**Comment by tsaridas** [comment]
iter_events is being used in reactor, cherrypy and raet so yes it would make sense to raise an exception and try to catch it in all places being used or all the above would have issues when master is restarted.

**Comment by thatch45** [comment]
My concern is to avoid changing behavior without explicitly adding the option. Many people use the event system api and I don't like pulling the rug out from under people unless we have to. Hence why I would rather make it an option.

**Comment by jeanpralo** [comment]
True, makes sense.


Well #37438 is fixing  so I guess we can close #37335  ?

**Comment by tsaridas** [comment]
verified fix for 2016.3.4 and get_event_noblock

**Comment by DmitryKuzmenko** [comment]
It looks I've done here.
